### ISSUE TITLE
"RuntimeException: Unexpected EOF while reading" when building

### ISSUE THREAD

**Issue by tgk** [pixie-lang/pixie#268]
I'm unable to build the [current HEAD](https://github.com/pixie-lang/pixie/commit/f76d32c0d272f7dcc98e68721adcf00d15c68fc0), both from my normal Mac OSX box and using Docker. The last lines of a `docker build .` produces the following.

```
./pixie-vm -c pixie/uv.pxi -c pixie/io.pxi -c pixie/stacklets.pxi -c pixie/stdlib.pxi
sh: 1: c++: not found
in internal function run_load_stdlib

in internal function load-ns

in internal function load-file

in internal function load-reader

in pixie function main

in internal function load-ns

in internal function load-file

in internal function load-reader

in pixie function with-config

in /usr/src/pixie/pixie/ffi-infer.pxi at 212:10
        ~(run-infer *config* @*bodies*))))
         ^
in pixie function run-infer

in /usr/src/pixie/pixie/ffi-infer.pxi at 196:16
        result (read-string (io/run-command cmd-str))
               ^
in internal function read-string

RuntimeException: Unexpected EOF while reading

[/usr/src/pixie /usr/src/pixie/lib /usr/src/pixie/include /usr/src/pixie/../lib /usr/src/pixie/../include]
c++ -O3 -pthread -fomit-frame-pointer -Wall -Wno-unused  /tmp/tmp.cpp  -I /usr/src/pixie  -I /usr/src/pixie/lib  -I /usr/src/pixie/include  -I /usr/src/pixie/../lib  -I /usr/src/pixie/../include   -o /tmp/a.out && /tmp/a.out
make[1]: Leaving directory '/usr/src/pixie'
 ---> 888d42776c9f
Removing intermediate container 1c5969f5bd3c
Step 4 : ENTRYPOINT /usr/bin/pxi
 ---> Running in ae445a116f6b
 ---> 920c222c2a70
Removing intermediate container ae445a116f6b
Successfully built 920c222c2a70
```

Naturally, the resulting image does not work (but the error message is slightly unexpected)

```
$ docker run --rm -it pixie
[/usr/src/pixie /usr/src/pixie/lib /usr/src/pixie/include /usr/src/pixie/../lib /usr/src/pixie/../include]
c++ -O3 -pthread -fomit-frame-pointer -Wall -Wno-unused  /tmp/tmp.cpp  -I /usr/src/pixie  -I /usr/src/pixie/lib  -I /usr/src/pixie/include  -I /usr/src/pixie/../lib  -I /usr/src/pixie/../include   -o /tmp/a.out && /tmp/a.out
sh: 1: c++: not found
in internal function run_load_stdlib

in internal function load-ns

in internal function load-file

in internal function load-reader

in pixie function main

in internal function load-ns

in internal function load-file

in internal function load-reader

in pixie function with-config

in /usr/src/pixie/pixie/ffi-infer.pxi at 212:10
        ~(run-infer *config* @*bodies*))))
         ^
in pixie function run-infer

in /usr/src/pixie/pixie/ffi-infer.pxi at 196:16
        result (read-string (io/run-command cmd-str))
               ^
in internal function read-string

RuntimeException: Unexpected EOF while reading
```

**Comment by heyLu** [comment]
Ah sorry, the Dockerfile hasn't been updated for a while. I'll try to have a look this weekend. Until then you'll have to use the official instructions.

**Comment by tgk** [comment]
Alright - still seems to break. I've checked out the `0.1` tag and it still fails with the same error, so I suspect something might be very odd on my setup. It's very puzzling that it fails on both my system and in Docker with the same error (Unexpected EOF while reading).

**Comment by tgk** [comment]
Pulling the official repo confirms that something on my system is odd:

```
$ docker run --rm -ti pixielang/pixie
Unable to find image 'pixielang/pixie:latest' locally
Pulling repository pixielang/pixie
508bd1e64c9d: Download complete
511136ea3c5a: Download complete
19df420c532f: Download complete
91f2410be649: Download complete
f77b7664312b: Download complete
f87cd6df6e91: Download complete
1898fb6f0ccb: Download complete
Status: Downloaded newer image for pixielang/pixie:latest
[/usr/src/pixie /usr/src/pixie/lib /usr/src/pixie/include /usr/src/pixie/../lib /usr/src/pixie/../include]
c++ -O3 -pthread -fomit-frame-pointer -Wall -Wno-unused  /tmp/tmp.cpp  -I /usr/src/pixie  -I /usr/src/pixie/lib  -I /usr/src/pixie/include  -I /usr/src/pixie/../lib  -I /usr/src/pixie/../include   -o /tmp/a.out && /tmp/a.out
sh: 1: c++: not found
in internal function run_load_stdlib

in internal function load-ns

in internal function load-file

in internal function load-reader

in pixie function main

in internal function load-ns

in internal function load-file

in internal function load-reader

in pixie function with-config

in /usr/src/pixie/pixie/ffi-infer.pxi at 212:10
        ~(run-infer *config* @*bodies*))))
         ^
in pixie function run-infer

in /usr/src/pixie/pixie/ffi-infer.pxi at 196:16
        result (read-string (io/run-command cmd-str))
               ^
in internal function read-string

RuntimeException: Unexpected EOF while reading
```

Versions are:

```
$ boot2docker -v
Boot2Docker-cli version: v1.4.1
Git commit: 43241cb
Usage: boot2docker [<options>] {help|init|up|ssh|save|down|poweroff|reset|restart|config|status|info|ip|shellinit|delete|download|upgrade|version} [<args>]
```

and

```
$ docker -v
Docker version 1.4.1, build 5bc2ff8
```

I'll bump and see if it helps.

**Comment by tgk** [comment]
Upgraded, but can confirm that error persists. New versions are

```
$ boot2docker -v
Boot2Docker-cli version: v1.5.0
Git commit: ccd9032
Usage: boot2docker [<options>] {help|init|up|ssh|save|down|poweroff|reset|restart|config|status|info|ip|shellinit|delete|download|upgrade|version} [<args>]

$ docker -v
Docker version 1.5.0, build a8a31ef
```

Again, I suspect something must be odd on my system as the error is present both when running on my system "native" and through Docker.

**Comment by halgari** [comment]
The EOF while reading error is a red herring. ffi-infer is trying to
compile .cpp a file, and run it. When it reads the output it expects to
find EDN data, but instead it's finding a bash error "c++ not found". We
somehow need to install c++ or link g++ to c++ or fix ffi-infer to
auto-configure to either g++ or c++.

The best thing about unix standards is, there are so many to choose from.

Timothy

On Fri, Apr 3, 2015 at 3:08 PM, Thomas G. Kristensen <
notifications@github.com> wrote:

> Upgraded, but can confirm that error persists. New versions are
>
> $ boot2docker -v
> Boot2Docker-cli version: v1.5.0
> Git commit: ccd9032
> Usage: boot2docker [<options>] {help|init|up|ssh|save|down|poweroff|reset|restart|config|status|info|ip|shellinit|delete|download|upgrade|version} [<args>]
>
> $ docker -v
> Docker version 1.5.0, build a8a31ef
>
> Again, I suspect something must be odd on my system as the error is
> present both when running on my system "native" and through Docker.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pixie-lang/pixie/issues/268#issuecomment-89420538>.
>



-- 
“One of the main causes of the fall of the Roman Empire was that–lacking
zero–they had no way to indicate successful termination of their C
programs.”
(Robert Firth)

**Comment by tgk** [comment]
It is indeed - mo standards, mo problems

Both `c++` and `g++` exist on my path so it's still a bit puzzling.

Also, thanks for all the help!

**Comment by thomasmulvaney** [comment]
@tgk, I've had issues with the published docker image too, but running `docker build .` in the Pixie dir makes an image that works for me.

**Comment by tgk** [comment]
@thomasmulvaney, oddly building it myself didn't  work either. I'm going to retry with tag `0.1`.

**Comment by tgk** [comment]
Nope, still no luck.

**Comment by thomasmulvaney** [comment]
Very strange. I had the exact same errors you were having when pulling the built image, but the local build seemed to work fine. The weirdest thing is that the docker build process installs in GCC which makes the c++ error puzzling.

**Comment by heyLu** [comment]
I am very confused why `docker build` would for @thomasmulvaney, as there is no `c++` installed in the docker image. I've added it in [a branch](https://github.com/heyLu/pixie/tree/fix-dockerfile), but now it fails with a different error message, which I've also seen when building outside of docker:

```
make compile_basics
make[1]: Entering directory '/home/lu/t/pixie/pixie'




WARNING: Compiling core libs. If you want to modify one of these files delete the .pxic files first




./pixie-vm -c pixie/uv.pxi -c pixie/io.pxi -c pixie/stacklets.pxi -c pixie/stdlib.pxi
in internal function run_load_stdlib

in internal function load-ns

in internal function load-file

in pixie function first

in pixie/stdlib.pxi at 643:8

       ^
RuntimeException: Var satisfies? is undefined

make[1]: Leaving directory '/home/lu/t/pixie/pixie'
```

**Comment by heyLu** [comment]
Ok, the error I posted above was due to stale files being present. You can avoid that by using `git archive master | docker build -` to build the docker image. (Running `make clean` might also help, but I'm not sure which files/commands were the problem.)

To be able to build pixie you need a c++ compiler and the boost libraries installed on your machine. I fixed that in the dockerfile in #273.
